File: kernel/aux.py
'''
Created on Apr 29, 2014

@author: Yair Daon. email: fisrtname.lsatname@gmail.com
Feel free to write to me about my code!
'''
from numpy.linalg import norm
from numpy import zeros
from numpy import array
from numpy import einsum
from numpy import ravel

from math import exp

def cov(x,y, *args):
    '''
    calculate autocovariance as
    k(x,y) = exp(   -||x-y||^2 / r   )
    :param x:
        first point in space
    :param y: 
        second point in space
    :param r:
        the characteristic length scale of
        of the covariance
    :param d:
        the covariance for two points that
        are arbitrarily far apart
        
    * ``cov`` - the covariance between x,y with 
        characteristic length r
    '''
    
    # unpack
    r = args[0]
    d = args[1]
    
    if len(args) > 2:
        A = args[2]
        t = x - y
        dist = einsum(' i, ij, j ', t, A, t)/r

    else:
        dist = norm(x-y)/r  
#     t = dist*1.7320508075688772 # sqrt(3)
# 
#     return (1 + t)*math.exp(-t)

    return  d*exp(  -(dist*dist)/2  ) 

def cov_mat(X, *args): 
    '''
    create and return the covariance matrix for the observations
    :param X: 
        a list of locations in space, for which we calculate covariance
    :param r:
        characteristic length scale
    :param d: the covariance for two points that
        are arbitrarily far apart
    '''
    
    #decide on the size of
    n = len(X)
    
    # allocate memory
    C = zeros( (n,n) )
    
    # set the values of the covariance, exploiting symmetry
    for i in range(n):
        for j in range (i):
            C[i,j] = cov(X[i],X[j], *args)
            C[j,i] = C[i,j]
        C[i,i] = cov(X[i], X[i], *args)
    return C

def cov_vec(X,w, *args):
    '''
    creates a vector of covariances, between 
    X(a list of numpy arrays) and w (a numpy
    array of same size).
    :param X: 
        a list of locations
    :param w:
        a specific location for which we calculate covariances
    :param r: 
        characteristic length scale
    :param d: the covariance for two points that
        are arbitrarily far apart
        
    returns a vector v s.t. v_i =cov(x_i,w)
    '''
    
    return array( 
                    [cov(x,w, *args) for x in X]
#                     map(lambda x: cov(x,w, *args) , X )
                    )

      
def tychonoff_solver( specs, b ):
    '''
    solve Ax = b  using tychonoff regularization or, 
    equvalently, multiply A^{-1}b stably.
    we return the solution to the optimization problem
    x = argmin ||Ax -b||^2 + reg*||x||^2.
    U, S, V is the SVD of A ( A = USV and V is NOT transposed!! )
    reg is the regularization coefficient
    :param U,S,V: the SVD of the matrix A. It holds that
        U*S*V = A (V and not V*, this is what python's
        numpy.linalg.svd(A) returns)
    :param b:
        the vector for which we solve
    * ``x`` - solution to the abovementioned optimization problem
    '''
    S = specs.S
    b = ravel(b)
    c = einsum( ' ij ,  i -> j' , specs.U, b)  # c = U^t * b
    y = c*S/(S*S + specs.reg )            # y_i  = s_i * c_i  /  (s_i^2 + reg
    x = einsum( 'ij , i -> j', specs.V ,  y )  # x = V^t * y

    return x

File: C/aux.c
#include "aux.h"
#include <math.h>
#include <stdlib.h>

double *solver(double *U, double *S, double *V, double *b , double reg, int n) {
	/*
	 * solve Ax = b using tychonoff regularization
	 * A = U * S * V (no transpose here!)
	 * U is an n by n matrices
	 * V is an n by n matrices
	 * S is an n by 1 vector
	 * b is an n by 1 vector
	 * reg is the regularization we use
	 */

	// allocate memory and initialize
	double *c = calloc((size_t)n , sizeof(double)); // also
	double *x = calloc((size_t)n , sizeof(double)); // sets
	double *y = calloc((size_t)n , sizeof(double)); // zeros
	int i,j;
	double tmp;

	//c = U^t * b
	for (i = 0 ; i < n ; i++) {
		tmp = b[i];
		for (j = 0 ; j < n ; j++ ) {
			c[j] += U[ i*n + j ]*tmp;
		}
	}

	// y_i  = s_i * c_i  /  (s_i^2 + reg
	for (i = 0 ; i < n ; i++) {
		tmp =S[i];
		y[i] = c[i]*tmp/( tmp*tmp + reg );
	}

	// x = V^t * y
	for (i = 0 ; i < n ; i++) {
			tmp = y[i];
			for (j = 0 ; j < n ; j++ ) {
				x[j] += V[ i*n + j ]*tmp;
			}
		}

	free(y);
	free(c);
	return x;

}


double *covMat(double *X, double r, double d, int nvecs, int veclen) {

	double result;
	double dist;
	int n;

	double *K;
	K = malloc(nvecs * nvecs * sizeof(double));

	int row, col;
	for( row = 0 ; row < nvecs ; row++) {
		for (col = 0 ; col < row+1 ; col++) {

			result = 0.0;
			for (n = 0; n < veclen; n++) {
					dist = X[col*veclen + n] - X[row*veclen + n];
					result += dist * dist;
			}

			result = d*exp(  -result/(2.0*r*r)  );

			// set the element of the cov matrix
			K[row*nvecs + col] = result;
			K[col*nvecs + row] = result;
		}
	}
	return K;
}

double *covVec(double *X, double *x, double r, double d, int nvecs, int veclen) {

	double result;
	double dist;
	int n;

	double *k;
	k = malloc(nvecs * sizeof(double));

	int row;
	for( row = 0 ; row < nvecs ; row++) {

		result = 0.0;
		for (n = 0; n < veclen; n++) {
				dist = x[n] - X[row*veclen + n];
				result += dist * dist;
		}

		result = d*exp(  -result/(2.0*r*r)  );
		k[row] = result;
	}
	return k;
}


double cov(double *x, double *y,  double r, double d, int N) {

	double result = 0.0;
	double dist;
	int n;
    	for (n = 0; n < N; n++) {
     	   	dist = y[n] - x[n];
		result += dist * dist;
    	}

	result = result /(2.0*r*r);
	result = d*exp(-result);
    	return result;
}
